File: apps/api/app/metrics.py
"""Server-side metric recomputation from raw trials — the source of truth.

Mirrors packages/core/src/metrics.ts exactly (PRD §5/§6). Client metrics are
advisory only.
"""

import statistics
from typing import Any

Trial = dict[str, Any]


def _clamp(v: float, lo: float = 0.0, hi: float = 1.0) -> float:
    return min(hi, max(lo, v))


def _scoreable(trials: list[Trial]) -> list[Trial]:
    return [t for t in trials if not t["false_start"] and not t["interrupted"]]


def _rts(trials: list[Trial]) -> list[float]:
    return [
        t["response_ms"] - t["onset_ms"]
        for t in trials
        if not t["false_start"]
        and not t["interrupted"]
        and t["response_ms"] is not None
        and t["correct"]
    ]


def median(xs: list[float]) -> float | None:
    return statistics.median(xs) if xs else None


def iqr(xs: list[float]) -> float | None:
    if len(xs) < 4:
        return None
    qs = statistics.quantiles(xs, n=4, method="inclusive")
    return qs[2] - qs[0]


def compute_metrics(game_id: str, trials: list[Trial]) -> dict[str, Any]:
    sc = _scoreable(trials)
    rt = _rts(trials)
    n_false = sum(1 for t in trials if t["false_start"])
    accuracy = (sum(1 for t in sc if t["correct"]) / len(sc)) if sc else 0.0
    med = median(rt)
    base = {
        "accuracy": accuracy,
        "median_rt_ms": med,
        "rt_iqr_ms": iqr(rt),
        "false_start_rate": (n_false / len(trials)) if trials else 0.0,
        "scoreable_trials": len(sc),
    }
    base["performance_index"] = performance_index(game_id, trials, base)
    return base


def performance_index(game_id: str, trials: list[Trial], m: dict[str, Any]) -> float:
    med = m["median_rt_ms"]
    sc = _scoreable(trials)

    if game_id == "flash_point":
        speed = 0.0 if med is None else _clamp((450 - med) / 250)
        return _clamp(0.7 * speed + 0.3 * (1 - m["false_start_rate"]))

    if game_id == "vector":
        speed = 0.0 if med is None else _clamp((900 - med) / 500)
        rev = [t for t in sc if t["payload"].get("reverse")]
        rev_acc = (sum(1 for t in rev if t["correct"]) / len(rev)) if rev else 0.0
        return _clamp(0.55 * m["accuracy"] + 0.35 * speed + 0.10 * rev_acc)

    if game_id == "stackwise":
        resp = [t for t in sc if t["payload"].get("is_match") is not None]
        matches = [t for t in resp if t["payload"]["is_match"]]
        non = [t for t in resp if not t["payload"]["is_match"]]
        hit = (sum(1 for t in matches if t["correct"]) / len(matches)) if matches else 0.0
        cr = (sum(1 for t in non if t["correct"]) / len(non)) if non else 0.0
        return _clamp((hit + cr) / 2)

    if game_id == "drift_watch":
        if not sc:
            return 0.0
        fr = [
            (t["payload"]["n_correct"] / len(t["payload"]["target_ids"]))
            if t["payload"].get("target_ids")
            else 0.0
            for t in sc
        ]
        return _clamp(sum(fr) / len(fr))

    if game_id == "wide_angle":
        if not sc:
            return 0.0
        c_acc = sum(1 for t in sc if t["payload"].get("center_ok")) / len(sc)
        b_acc = sum(1 for t in sc if t["payload"].get("bearing_ok")) / len(sc)
        return _clamp(0.5 * c_acc + 0.5 * b_acc)

    if game_id == "echo_grid":
        if not sc:
            return 0.0
        js = []
        for t in sc:
            p = t["payload"]
            inter = p["n_correct"]
            union = p["n_cells"] + len(p["chosen"]) - inter
            js.append(inter / union if union else 0.0)
        return _clamp(sum(js) / len(js))

    raise ValueError(f"Unknown game_id: {game_id}")


def display_score(performance_index_value: float, difficulty: int) -> int:
    """PRD §9: round(1000 · performance_index · (0.5 + difficulty/200))."""
    return round(1000 * performance_index_value * (0.5 + difficulty / 200))

File: packages/core/src/metrics.ts
/** Session metrics + per-game performance index — PRD §5/§6.
 * Implemented identically server-side (Python) as truth; this copy powers
 * instant client display and the bot/staircase tests. */

import type { GameId, SessionMetrics, TrialResult } from "./types";

const clamp = (v: number, lo: number, hi: number) => Math.min(hi, Math.max(lo, v));

export function median(xs: number[]): number | null {
  if (xs.length === 0) return null;
  const s = xs.slice().sort((a, b) => a - b);
  const mid = Math.floor(s.length / 2);
  return s.length % 2 ? s[mid]! : (s[mid - 1]! + s[mid]!) / 2;
}

export function iqr(xs: number[]): number | null {
  if (xs.length < 4) return null;
  const s = xs.slice().sort((a, b) => a - b);
  const q = (p: number) => {
    const idx = p * (s.length - 1);
    const lo = Math.floor(idx);
    const frac = idx - lo;
    return s[lo]! + (s[Math.min(lo + 1, s.length - 1)]! - s[lo]!) * frac;
  };
  return q(0.75) - q(0.25);
}

function rts(trials: TrialResult[]): number[] {
  return trials
    .filter((t) => !t.false_start && !t.interrupted && t.response_ms !== null && t.correct)
    .map((t) => t.response_ms! - t.onset_ms);
}

function scoreable(trials: TrialResult[]): TrialResult[] {
  return trials.filter((t) => !t.false_start && !t.interrupted);
}

export function computeMetrics(gameId: GameId, trials: TrialResult[]): SessionMetrics {
  const sc = scoreable(trials);
  const rt = rts(trials);
  const falseStarts = trials.filter((t) => t.false_start).length;
  const accuracy = sc.length ? sc.filter((t) => t.correct).length / sc.length : 0;
  const med = median(rt);
  const base: Omit<SessionMetrics, "performance_index"> = {
    accuracy,
    median_rt_ms: med,
    rt_iqr_ms: iqr(rt),
    false_start_rate: trials.length ? falseStarts / trials.length : 0,
    scoreable_trials: sc.length,
  };
  return { ...base, performance_index: performanceIndex(gameId, trials, base) };
}

export function performanceIndex(
  gameId: GameId,
  trials: TrialResult[],
  m: Omit<SessionMetrics, "performance_index">,
): number {
  const med = m.median_rt_ms;
  switch (gameId) {
    case "flash_point": {
      // 0.7·speed + 0.3·(1 − false_start_rate); speed = clamp((450 − medRT)/250)
      const speed = med === null ? 0 : clamp((450 - med) / 250, 0, 1);
      return clamp(0.7 * speed + 0.3 * (1 - m.false_start_rate), 0, 1);
    }
    case "reflex_drop": {
      // 0.5·speed + 0.35·accuracy + 0.15·(1 − false_start_rate).
      // speed anchors to six-choice RT (700 ms → 0, 400 ms → 1), not the simple-RT
      // scale Flash Point uses — picking a rod costs a Hick's-law step.
      const speed = med === null ? 0 : clamp((700 - med) / 300, 0, 1);
      return clamp(0.5 * speed + 0.35 * m.accuracy + 0.15 * (1 - m.false_start_rate), 0, 1);
    }
    case "vector": {
      // 0.55·acc + 0.35·speed + 0.10·reverse_accuracy; speed = clamp((900 − medRT)/500)
      const speed = med === null ? 0 : clamp((900 - med) / 500, 0, 1);
      const rev = scoreable(trials).filter((t) => (t.payload as { reverse?: boolean }).reverse);
      const revAcc = rev.length ? rev.filter((t) => t.correct).length / rev.length : 0;
      return clamp(0.55 * m.accuracy + 0.35 * speed + 0.1 * revAcc, 0, 1);
    }
    case "stackwise": {
      // balanced accuracy over responded presentations (is_match != null)
      const resp = scoreable(trials).filter(
        (t) => (t.payload as { is_match?: boolean | null }).is_match !== null &&
               (t.payload as { is_match?: boolean | null }).is_match !== undefined,
      );
      const matches = resp.filter((t) => (t.payload as { is_match: boolean }).is_match);
      const nonMatches = resp.filter((t) => !(t.payload as { is_match: boolean }).is_match);
      const hitRate = matches.length ? matches.filter((t) => t.correct).length / matches.length : 0;
      const crRate = nonMatches.length
        ? nonMatches.filter((t) => t.correct).length / nonMatches.length
        : 0;
      return clamp((hitRate + crRate) / 2, 0, 1);
    }
    case "drift_watch": {
      // mean(n_correct / n_targets) across rounds
      const rounds = scoreable(trials);
      if (!rounds.length) return 0;
      const fr = rounds.map((t) => {
        const p = t.payload as { n_correct: number; target_ids: number[] };
        return p.target_ids.length ? p.n_correct / p.target_ids.length : 0;
      });
      return clamp(fr.reduce((a, b) => a + b, 0) / fr.length, 0, 1);
    }
    case "wide_angle": {
      const sc = scoreable(trials);
      if (!sc.length) return 0;
      const cAcc = sc.filter((t) => (t.payload as { center_ok: boolean }).center_ok).length / sc.length;
      const bAcc = sc.filter((t) => (t.payload as { bearing_ok: boolean }).bearing_ok).length / sc.length;
      return clamp(0.5 * cAcc + 0.5 * bAcc, 0, 1);
    }
    case "echo_grid": {
      // mean Jaccard |chosen ∩ cells| / |chosen ∪ cells|
      const sc = scoreable(trials);
      if (!sc.length) return 0;
      const js = sc.map((t) => {
        const p = t.payload as { chosen: number[]; n_cells: number; n_correct: number; n_extra: number };
        const inter = p.n_correct;
        const union = p.n_cells + p.chosen.length - inter;
        return union ? inter / union : 0;
      });
      return clamp(js.reduce((a, b) => a + b, 0) / js.length, 0, 1);
    }
  }
}

/** Display score per session — PRD §9. */
export function displayScore(performance_index: number, difficulty: number): number {
  return Math.round(1000 * performance_index * (0.5 + difficulty / 200));
}
